contrib/babelfishpg_tsql/src/pltsql_utils.c

```diff
@@ -1153,4 +1156,226 @@ bool is_schema_from_db(Oid schema_oid, Oid db_id)
+
+			pltsql_update_target_table_alias(&(stmt->relation), stmt->fromClause);
+			if (stmt->withClause)
+				pltsql_cte_update_target_table_alias(stmt->withClause);
+		}
+	}	
+}
+
+/*
+ * If a table alias is used when specifying the target table, we need to refer to the
+ * FROM clause for table reference.
+ * Example:
+ * UPDATE tt SET tt.a = 1 FROM t AS tt WHERE tt.b = 2
+ *
+ * We also need to refer to the FROM clause for schema info.
+ * Example:
+ * UPDATE t SET t.a = 1 FROM sch.t WHERE t.b = 2
+ */
+void
+pltsql_update_target_table_alias(RangeVar **target_table, List *fromClause)
+{
+	ListCell *lc;
+
+	if (!(*target_table) || !fromClause || !IsA(*target_table, RangeVar) || (*target_table)->alias)
+		return;
+	
+	/*
+	 * For each table reference in fromClause, check if the table name or table alias
+	 * name matches the target table.
+	 * If yes, substitute target table with the table reference.
+	 */
+	foreach(lc, fromClause)
+	{
+		Node *n = lfirst(lc);
+		if (find_target_alias_in_tableref(target_table, n))
+			return;
+	}
+}
+
+static bool
+find_target_alias_in_tableref(RangeVar **target, Node *tblref)
+{
+	/* 
+ 	 * If the table refenrence is a JoinExpr, recursively check the 
+ 	 * join tree's left child and right child.
+ 	 */
+	if (IsA(tblref, JoinExpr))
+	{
+		JoinExpr *je = (JoinExpr *) tblref;
+		if (find_target_alias_in_tableref(target, (Node *) je->larg) ||
+			find_target_alias_in_tableref(target, (Node *) je->rarg))
+			return true;
+	}
+	/*
+	 * If the table reference is an actual table (RangeVar), check if
+	 * the table name or table alias is the same as the target table name.
+	 * If yes, substitute the target table with the RangeVar.
+	 */
+	else if (IsA(tblref, RangeVar))
+	{
+		RangeVar *rv = (RangeVar *) tblref;
+		if (pg_strcasecmp((*target)->relname, rv->relname) == 0)
+		{
+			if ((*target)->schemaname && 
+				(!rv->schemaname || pg_strcasecmp((*target)->schemaname, rv->schemaname) != 0))
+			{
+				ereport(ERROR,
+						(errcode(ERRCODE_SYNTAX_ERROR),
+						 errmsg("The objects \"%s.%s\" and \"%s\" in the FROM clause have the same exposed names. " \
+								"Use correlation names to distinguish them.", 
+								(*target)->schemaname, (*target)->relname, rv->relname)));
+			}
+
+			*target = rv;
+			return true;
+		}
+		else if (rv->alias && pg_strcasecmp((*target)->relname, rv->alias->aliasname) == 0)
+		{
+			if ((*target)->schemaname)
+				ereport(ERROR,
+						(errcode(ERRCODE_SYNTAX_ERROR),
+						 errmsg("The correlation name \'%s\' has the same exposed name as table \'%s.%s\'.", 
+								rv->alias->aliasname, (*target)->schemaname, (*target)->relname)));
+			*target = rv;
+			return true;
+		}
+	}
+	/* 
+ 	 * Currently we only consider RangeVar and JoinExpr cases. In the
+ 	 * future, if there are concrete use cases, we'll add support for 
+ 	 * more table reference types
+ 	 */
+	return false;
+}
+
+void
+pltsql_update_table_reference(RangeVar *target_table, List *fromClause, ParseState *pstate)
```

Comment:
Then instead of pre_parse_analyze and post_set_target_table, we would need to add a hook around setTargetTable. 
```
if (hook)
   /* 
    * if duplication exists, call setTargetTable(alsosouorce=false) 
    * and reset r_table
    */
   qry->resultRelation = (*hook) () 
else
   qry->resultRelation = setTargetTable()
```
Yes this might be a better way, and this also avoid dealing with CTEs